core/Db/Adapter/Mysqli.php

```diff
@@ -9,17 +9,19 @@
 namespace Piwik\Db\Adapter;
 
 use Exception;
-use Piwik\Config;
 use Piwik\Db;
+use Piwik\Db\Adapter\MysqlTrait;
 use Piwik\Db\AdapterInterface;
-use Piwik\Piwik;
 use Zend_Config;
 use Zend_Db_Adapter_Mysqli;
 
 /**
+ * Mysqli adapter
  */
 class Mysqli extends Zend_Db_Adapter_Mysqli implements AdapterInterface
 {
+    use MysqlTrait;
```

Comment:
I don't like the use of a trait in this PR, where it seems to be just a place to gather an assortment of methods two classes have in common. It violates the single responsibility principle and is thus hard to understand exactly what it represents (evidenced by the generic name, "MysqlTrait").

@tsteur @sgiehl  do you have the same feeling I do? or are you guys ok w/ this use of a trait?

Reply:
Perhaps `MySqlAdapterTrait` would be a better name but this trait is how traits were meant to be used. It also fulfills the DRY principle.

>Traits are a mechanism for code reuse in single inheritance languages such as PHP. A Trait is intended to reduce some limitations of single inheritance by enabling a developer to reuse sets of methods freely in several independent classes living in different class hierarchies.

Piwik\Db\Adapter\Mysqli extends Zend_Db_Adapter_Mysqli while Piwik\Db\Adapter\Pdo\Mysql extends Zend_Db_Adapter_Pdo_Mysql.

Reply:
I understand the technical purpose of traits and of inheritance, but using them just to separate out common code is (for lack of a better term) lower order design. It results in code that doesn't have a well defined purpose, and thus loses reusability and coherence (there's a reason "composition over inheritance" is a thing, it forces you to think further than DRY). I've made this mistake before, one good example being: https://github.com/matomo-org/matomo/blob/3.x-dev/core/BaseFactory.php. This class had limited use at the time, no use now, and is far less powerful than DI.

The point I'm trying to make isn't that we shouldn't use traits, but that small refactors like this done simply to reduce the amount of code have a tendency to become cruft in the future. I think ideally, if we want to reduce the redundancy in these two classes, we'd think about it in a separate issue/PR rather than attach it to a bug fix for expedience.

In this specific PR, the `hasBlobDataType()` and related methods don't really need to be moved, they're simple enough that removing the redundancy won't make the code easier to understand (in fact I would say it would do the opposite, since now you can't see where those methods are implemented at a glance). What's important from a code coherence standpoint is separating & reusing the logic to detect & handle azure, into some unit that exists specifically to encapsulate azure-mysql logic. I hope that makes sense. (And I don't necessarily speak for other devs.)